### PR TITLE
[WIP][OSPP][Java SDK] Java SDK Async API

### PR DESCRIPTION
**Pull request by yyy1000**
I have finished the Async API code, doc, and unit tests.
It also includes the Java SDK code rebased changes from [kenvi](https://github.com/kenvifire/WasmEdge).
The roadmap is in [Kenvi's PR](https://github.com/WasmEdge/WasmEdge/pull/435)

### CODE REVIEW

**Comment by hydai** [comment]
Hi @yyy1000 
We already merged @kenvifire 's PR. Please rebase to the latest master.

**Comment by dannypsnl** [comment]
The PR of Java SDK gets merged, would you mind rebasing on the new changes? I would start reviewing this PR recently.

**Comment by yyy1000** [comment]
> I would start reviewing this PR recently, and can I close #1873?

Yes, you can close it.

**Comment by yyy1000** [comment]
> Hi @yyy1000 We already merged @kenvifire 's PR. Please rebase to the latest master.

Hi, @kenvifire and I have rebased the latest master, you can continue to review this PR. Maybe you can merge before the end of this month (to meet the deadline for this activity) and I will continue to contribute to this Java SDK. 😀

**Comment by hydai** [comment]
Hi @yyy1000 I believe the rebase is totally wrong. There are so many duplicated commits. Because we squashed the commits in the previous PR, you may need to move your commits on the current master.

**Comment by kenvifire** [comment]
@hydai 

> Hi @yyy1000 I believe the rebase is totally wrong. There are so many duplicated commits. Because we squashed the commits in the previous PR, you may need to move your commits on the current master.

My bad, I didn't notice that you squashed the commits. Removed duplicates now, please take a look. Thanks.

**Comment by kenvifire** [comment]
By the way, I turned off macOS test in this MR. It's been flaky recently on CI and never happens on my Mac, and I'll continue investigating it.

File: bindings/java/wasmedge-jni/lib/WasmEdgeVM.c [hunk 1]
```diff
@@ -597,4 +598,251 @@ JNIEXPORT void JNICALL Java_org_wasmedge_WasmEdgeVM_registerModuleFromImport(
...
+      break;
+    }
+    wasm_params[i] = val;
+  }
+
+  WasmEdge_Async *async = WasmEdge_VMAsyncExecuteRegistered(
+      vmContext, wModName, wFuncName, wasm_params, paramLen);
+
+  return createJAsyncObject(env, async);
 }
```
Comment (dannypsnl) [inline bindings/java/wasmedge-jni/lib/WasmEdgeVM.c:848]: misc: eof

**Review by dannypsnl — CHANGES_REQUESTED** [review]

**Comment by dannypsnl** [comment]
@kenvifire Would you like to provide your macOS machine information? Like system version & CPU.

Another thing is we have the head format for `*.c/cpp`, would you mind adding them?
```
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2019-2022 Second State INC
```

**Comment by kenvifire** [comment]
> @kenvifire Would you like to provide your macOS machine information? Like system version & CPU.
> 
> 
> 
> Another thing is we have the head format for `*.c/cpp`, would you mind adding them?
> 
> ```
> 
> // SPDX-License-Identifier: Apache-2.0
> 
> // SPDX-FileCopyrightText: 2019-2022 Second State INC
> 
> ```

I can create a separate pull request for this. However, yyy1000 is doing this feature under an open source activity and the deadline is this Sunday. Would you mind merge this PR
first?

**Comment by hydai** [comment]
Sure, I think we can merge it first and fix these minor issues later.

**Review by hydai — APPROVED** [review]

**Pull request merged**